cirq/qis/measures.py

```diff
@@ -38,36 +56,211 @@ def fidelity(state1: np.ndarray, state2: np.ndarray) -> float:
+    if isinstance(state1, int) and isinstance(state2, int):
+        if validate and qid_shape is not None:
+            dim = np.prod(qid_shape)
+            if state1 >= dim:
+                raise ValueError(
+                    'Invalid state for given qid shape: '
+                    'The maximum computational basis state for qid shape '
+                    f'{qid_shape} is {dim - 1}, but {state1} was given.'
+                )
+            if state2 >= dim:
+                raise ValueError(
+                    'Invalid state for given qid shape: '
+                    'The maximum computational basis state for qid shape '
+                    f'{qid_shape} is {dim - 1}, but {state2} was given.'
+                )
+        return float(state1 == state2)
+
+    # Two ProductStates
+    if isinstance(state1, value.ProductState) and isinstance(state2, value.ProductState):
+        if len(state1) != len(state2):
+            raise ValueError(
+                'Mismatched number of qubits in product states: '
+                f'{len(state1)} and {len(state2)}.'
+            )
+        if validate and qid_shape is not None and qid_shape != (2,) * len(state1):
+            raise ValueError(
+                'Invalid state for given qid shape: '
+                f'Specified shape {qid_shape} but product state '
+                f'has shape {(2,) * len(state1)}.'
+            )
+        prod = 1.0
+        for q, s1 in state1:
+            s2 = state2[q]
+            prod *= np.abs(np.vdot(s1.state_vector(), s2.state_vector()))
+        return prod ** 2
+
+    # Two numpy arrays that are either state vector, state tensor, or
+    # density matrix
+    if (
+        isinstance(state1, np.ndarray)
+        and state1.dtype.kind == 'c'
+        and isinstance(state2, np.ndarray)
+        and state2.dtype.kind == 'c'
+    ):
+        state1, state2 = _numpy_arrays_to_state_vectors_or_density_matrices(
+            state1, state2, qid_shape=qid_shape, validate=validate, atol=atol
+        )
+        return _fidelity_state_vectors_or_density_matrices(state1, state2)
+
+    # Use QuantumState machinery for the general case
+    if qid_shape is None:
+        qid_shape = infer_qid_shape(state1, state2)
+    state1 = quantum_state(state1, qid_shape=qid_shape, validate=validate, atol=atol)
+    state2 = quantum_state(state2, qid_shape=qid_shape, validate=validate, atol=atol)
+    state1 = state1.density_matrix() if state1._is_density_matrix() else state1.state_vector()
+    state2 = state2.density_matrix() if state2._is_density_matrix() else state2.state_vector()
+    return _fidelity_state_vectors_or_density_matrices(state1, state2)
+
+
+def _numpy_arrays_to_state_vectors_or_density_matrices(
+    state1: np.ndarray,
+    state2: np.ndarray,
+    qid_shape: Optional[Tuple[int, ...]],
+    validate: bool,
+    atol: float,
+) -> Tuple[np.ndarray, np.ndarray]:
+    if state1.ndim > 2 or state1.ndim == 2 and state1.shape[0] != state1.shape[1]:
```

Comment:
add parens to make the clauses a bit clearer - I had to look twice ("if state1.dim is greater or equal to 2, why not use >=...?:)")

Reply:
Done.